src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,514 @@
+        /// </summary>
+        public int Count => this.nodes.Count;
+
+        /// <summary>
+        /// Gets the priority comparer of the priority queue.
+        /// </summary>
+        public IComparer<TPriority> Comparer { get; }
+
+        /// <summary>
+        /// Enqueues the specified element and associates it with the specified priority.
+        /// </summary>
+        public void Enqueue(TElement element, TPriority priority)
+        {
+            if (element is null)
+            {
+                throw new ArgumentNullException(nameof(element));
+            }
+
+            if (priority is null)
+            {
+                throw new ArgumentNullException(nameof(priority));
+            }
+
+            // Add the node at the end
+            var node = (element, priority);
+            this.nodes.Add(node);
+
+            // Restore the heap order
+            var lastNodeIndex = this.GetLastNodeIndex();
+            this.MoveUp(node, lastNodeIndex);
+        }
+
+        /// <summary>
+        /// Gets the element associated with the minimal priority.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The queue is empty.</exception>
+        public TElement Peek()
+        {
+            if (this.TryPeek(out var element, out var priority))
+            {
+                return element;
+            }
+            else
+            {
+                throw new InvalidOperationException(
+                    "The priority queue is empty, cannot get the element with minimal priority.");
+            }
+        }
+
+        /// <summary>
+        /// Dequeues the element associated with the minimal priority.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The queue is empty.</exception>
+        public TElement Dequeue()
+        {
+            if (this.TryDequeue(out var element, out var priority))
+            {
+                return element;
+            }
+            else
+            {
+                throw new InvalidOperationException(
+                    "The priority queue is empty, cannot dequeue an element.");
+            }
+        }
+
+        /// <summary>
+        /// Dequeues the element associated with the minimal priority
+        /// </summary>
+        /// <returns>
+        /// <see langword="true"/> if the priority queue is non-empty; <see langword="false"/> otherwise.
+        /// </returns>
+        public bool TryDequeue([MaybeNullWhen(false)] out TElement element, [MaybeNullWhen(false)] out TPriority priority)
+        {
+            if (this.nodes.Count == 0)
+            {
+                element = default;
+                priority = default;
+                return false;
+            }
+            else
+            {
+                (element, priority) = this.nodes[RootIndex];
+                this.Remove(RootIndex);
+                return true;
+            }
```

Comment:
Nit: Try methods in this library are generally implemented with the positive case first, e.g.
```C#
if (nodes.Count != 0)
{
    (element, priority) = this.nodes[RootIndex];
    Remove(RootIndex);
    return true;
}

element = default;
priority = default;
return false;
```